url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/FuzzySystem.java

```diff
@@ -0,0 +1,938 @@
+		
+		// Average the fuzzification of the input set 
+		for(int i = 0; i < vels.length; i ++){
+			
+			ppp += this.inverseGrade(vels[i], 13, 18);
+			pp 	+= this.trapezoid(vels[i], 12, 17, 29, 34); 
+			p 	+= this.trapezoid(vels[i], 28, 33, 45, 50); 
+			mp  += this.trapezoid(vels[i], 44, 49, 61, 66); 
+			mf  += this.trapezoid(vels[i], 60, 65, 77, 82); 
+			f   += this.trapezoid(vels[i], 76, 81, 93, 98); 
+			ff  += this.trapezoid(vels[i], 92, 97, 109, 114);
+			fff += this.grade(vels[i], 108, 113);
+			
+		}
+		ppp = ppp/vels.length;
+		pp 	= pp/vels.length; 
+		p   = p/vels.length; 
+		mp  = mp/vels.length; 
+		mf  = mf/vels.length; 
+		f   = f/vels.length; 
+		ff  = ff/vels.length;
+		fff = fff/vels.length;
+		
+		// No rules are needed to determine the output as there 
+		// is only a single set of membership functions, so we
+		// can now defuzzify. 
+		
+		double crisp = 0;
+		
+		double [] u = {ppp,pp,p,mp,mf,f,ff,fff};
+		double [] x   = {4,8,12,16,20,24,28,32}; 
+		
+		// First check to see if there is an obvious peak in the results of the rule
+		// evaluations,If so then use then use the maximum method else use the singleton method. 
+		if(this.defuzz_Max(u, x) > -1){
+			crisp = this.defuzz_Max(u, x);
+		}
+		else{
+			crisp = (int) this.defuzz_Singleton(u, x);
+		}
+		
+		// As this is defuzzifing into classifications a series of if
+		// then conditionals outlines the ranges of each type
+		if(crisp > 0 && crisp <= 4)
+			d = Dynamic.PIANISSIMO_POSSIBLE;
+		else if(crisp > 4 && crisp <= 8)
+			d = Dynamic.PIANISSIMO;
+		else if(crisp > 8 && crisp <= 12)
+			d = Dynamic.PIANO;
+		else if(crisp > 12 && crisp <= 16)
+			d = Dynamic.MEZZO_PIANO;
+		else if(crisp > 16 && crisp <= 20)
+			d = Dynamic.MEZZO_FORTE;
+		else if(crisp > 20 && crisp <= 24)
+			d = Dynamic.FORTE;
+		else if(crisp > 24 && crisp <= 28)
+			d = Dynamic.FORTISSIMO;
+		else if(crisp > 28 && crisp <= 32)
+			d = Dynamic.FORTISSIMO_POSSIBLE;
+			
+		return d;
+	}
+	private DesiredLoudness determineDynamics(int [] vels){
+		
+		return new DesiredLoudness(this.determineDynamic(vels), this.determineDynamicShape(vels));
+		
+	}
+	
+	// ==========================================================
+	// =		METHODS TO DETERMINE TARGET FEEL				=
+	// ==========================================================
+
+	// Converts note values (eg 4n for crotchets) into its
+	// millisecond value for the current bpm.
+	public double convertNVToMillis(String s, double bpm){
+		
+		double millisVal = 0;
+		double decimal = 0;
+		String [] stringVals = {"1n","2nd","2n","4nd","2nt",
+								"4nswg","4n","8nd","4nt","8nswg",
+								"8n","16nd","8nt","16nswg","16n",
+								"32nd","16nt","32n","32nt"};
+		double [] decimalValues ={ 1, 1.33333333, 2, 2.666666666, 3,
+								  3.2, 4, 5.333333333, 6, 6.4, 8, 
+								  10.666666667, 12, 12.8, 16, 
+								  21.333333333, 24, 32, 48};
+		
+		for(int i = 0; i < stringVals.length; i++){
+			if(s == stringVals[i]){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.